### PR TITLE
boards: gd32e507z_eval: add initial support

### PR DESCRIPTION
**Pull request by gmarull**
Add initial support for the GD32E507Z-EVAL board. Only basic peripherals
are enabled for now:

- UART
- GPIO (LEDs, keys)
- PWM

Others, e.g. I2C, EEPROM, DAC, etc. require further testing and likely
some driver adjustments.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

### CODE REVIEW

**Review by str4t0m — DISMISSED** [review]
Can't test on the board itself, but all looks good.

File: boards/arm/gd32e507z_eval/doc/index.rst [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+Before programming your board make sure to configure boot jumpers as
+follows:
+
+- JP3/4: Select 2-3 for both (boot from user memory)
+
+Using J-Link
+=============
+
+While the board comes with an embedded GD-Link programmer, it is not yet
+supported by OpenOCD. However, J-Link can be used to program the board using
```
[the remark this answers is not in the document]
Comment (gmarull) [reply boards/arm/gd32e507z_eval/doc/index.rst:96]: I haven't tried, will check

[the remark this answers is not in the document]
Comment (gmarull) [reply boards/arm/gd32e507z_eval/doc/index.rst]: indeed, it works, thanks for the suggestion!

**Review by str4t0m — DISMISSED** [review]

**Comment by nandojve** [comment]
Probably not related with PR but I couldn't test PyOCD. Any tips?

My device:

```
[26306.414466] usb 1-9: new full-speed USB device number 17 using xhci_hcd
[26306.563427] usb 1-9: New USB device found, idVendor=1a86, idProduct=7523, bcdDevice= 2.64
[26306.563428] usb 1-9: New USB device strings: Mfr=0, Product=2, SerialNumber=0
[26306.563429] usb 1-9: Product: USB Serial
[26306.566317] ch341 1-9:1.0: ch341-uart converter detected
[26306.567208] usb 1-9: ch341-uart converter now attached to ttyUSB0
[26306.842462] usb 1-7: new full-speed USB device number 18 using xhci_hcd
[26306.991243] usb 1-7: New USB device found, idVendor=28e9, idProduct=058f, bcdDevice= 1.00
[26306.991244] usb 1-7: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[26306.991245] usb 1-7: Product: CMSIS-DAP
[26306.991246] usb 1-7: Manufacturer: GD32 ARM

```

Check probe ID:
```
$ python3 -m pyocd list
  #   Probe                Unique ID                         
-------------------------------------------------------------
  0   GD32 ARM CMSIS-DAP   7CR7UIQJEENC2BODXRBZYTXCB3AWSK4M 
```

Try to flash:
```
$ west flash
-- west flash: rebuilding
ninja: no work to do.
-- west flash: using runner pyocd
-- runners.pyocd: Flashing file: /home/gfbudke/zephyr/mainline/zephyr/build/zephyr/zephyr.hex
0001377:CRITICAL:__main__:Device 7CR7UIQJEENC2BODXRBZYTXCB3AWSK4M not found
Traceback (most recent call last):
  File "/home/gfbudke/.local/lib/python3.8/site-packages/pyocd/probe/cmsis_dap_probe.py", line 174, in open
    self._link.open()
  File "/home/gfbudke/.local/lib/python3.8/site-packages/pyocd/utility/concurrency.py", line 28, in _locking
    return func(self, *args, **kwargs)
  File "/home/gfbudke/.local/lib/python3.8/site-packages/pyocd/probe/pydapaccess/dap_access_cmsis_dap.py", line 658, in open
    self._interface.open()
  File "/home/gfbudke/.local/lib/python3.8/site-packages/pyocd/probe/pydapaccess/interface/pyusb_backend.py", line 73, in open
    raise DAPAccessIntf.DeviceError("Device %s not found" % self.serial_number)
pyocd.probe.pydapaccess.dap_access_api.DAPAccessIntf.DeviceError: Device 7CR7UIQJEENC2BODXRBZYTXCB3AWSK4M not found

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/gfbudke/.local/lib/python3.8/site-packages/pyocd/__main__.py", line 150, in run
    status = cmd.invoke()
  File "/home/gfbudke/.local/lib/python3.8/site-packages/pyocd/subcommands/load_cmd.py", line 96, in invoke
    with session:
  File "/home/gfbudke/.local/lib/python3.8/site-packages/pyocd/core/session.py", line 341, in __enter__
    self.open()
  File "/home/gfbudke/.local/lib/python3.8/site-packages/pyocd/core/session.py", line 455, in open
    self._probe.open()
  File "/home/gfbudke/.local/lib/python3.8/site-packages/pyocd/probe/cmsis_dap_probe.py", line 197, in open
    raise self._convert_exception(exc) from exc
pyocd.core.exceptions.ProbeError: Device 7CR7UIQJEENC2BODXRBZYTXCB3AWSK4M not found
FATAL ERROR: command exited with status 1: pyocd flash -e sector -a 0x8000000 -t gd32e507ze /home/gfbudke/zephyr/mainline/zephyr/build/zephyr/zephyr.hex
```

**Comment by gmarull** [comment]
@nandojve looks like you don't have the CMSIS-Pack correctly installed. I can flash it:

```
0001187 I Loading /home/gmarull/ws/teslabs/zephyrproject/zephyr/build/zephyr/zephyr.hex at 0x08000000 [load_cmd]
[==================================================] 100%
0002900 I Erased 8192 bytes (1 sector), programmed 6144 bytes (6 pages), skipped 8192 bytes (8 pages) at 8.35 kB/s [loader]
```

**Review by nandojve — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
@gmarull can you rebase?

**Comment by gmarull** [comment]
> @gmarull can you rebase?

done

**Review by str4t0m — APPROVED** [review]
Failed tests are unrelated (qemu_x86_64: kernel, qemu_x86: logging ).

**Review by nandojve — APPROVED** [review]

**Review by soburi — APPROVED** [review]
LGTM

**Comment by stephanosio** [comment]
Please rebase to fix CI failure

**Pull request merged**